docs(feature): job prioritization (#8935)
camunda/camunda-docs: docs/components/concepts/job-workers.md

If it's not present in the gateway as a client stream, restart your worker. If it's not present as a consumer in one of the brokers, this indicates a bug. As a workaround, restart your gateway, which will cause some interruption in your service, but will force all streams for this gateway to be recreated properly.

## Job prioritization

Camunda supports job prioritization for pull-based job activation.

You can define job priority:

- On the process as a default.
- On supported job-creating tasks as an override, including service tasks, send tasks, and script or business rule tasks implemented as job workers.

> Job priority affects the order in which jobs are activated for workers. It does not change retries, failures, or completion semantics.

Reviewer: opinion: I would move this to the top of this section as it explains what job prioritisation is about before explaining how it can be configured. However, if this is the pattern we typically use feel free to keep it as it is.

Important behavior:

- Higher-priority jobs are activated before lower-priority jobs within a partition.

> - Ordering is best effort across the cluster, not a global guarantee across all partitions.

Reviewer: How technical do we typically get in these `docs/components/concepts/` pages? Should we add some technical information here to explain this behavior?

Author: @georgios-goulos we typically don't go very in-depth. I'll expand this bullet though

- If no priority is defined, the default priority is `0`.
- Task-level priority overrides process-level priority.
- Jobs with the same priority are activated by ascending job key.
- Low-priority jobs can starve under sustained high-priority load and may never be activated while workers remain fully occupied by higher-priority jobs.
- Camunda does not provide fairness, aging, or starvation mitigation in the engine.

If starvation becomes a problem, increase worker capacity or revise your priority assignments.

Priority values can be static integers or FEEL expressions. FEEL expressions are evaluated when the job is created, and the resulting integer is stored on the job.

Priority uses signed 32-bit integer semantics. Do not assume a fixed `0-99` engine limit.

:::note Compatibility for pre-8.10 jobs
Jobs created before 8.10 may not have a stored priority value in secondary storage.

> For compatibility, Camunda treats these jobs as if they had the default priority value of `0`.

Reviewer: We should clarify that when users filter/sort using the `priority` field in the relevant job APIs, the pre-8.10 jobs will be filtered out as the searching operation is using the secondary storage and we will not be performing a backfill for that so the `priority` field will not be populated for pre-8.10 jobs. However, the Zeebe engine will still treat such jobs as if they had a priority of 0.

Author: @georgios-goulos I’ll update the compatibility note 👌

If it's not present in the gateway as a client stream, restart your worker. If it's not present as a consumer in one of the brokers, this indicates a bug. As a workaround, restart your gateway, which will cause some interruption in your service, but will force all streams for this gateway to be recreated properly.

## Job prioritization

> Camunda supports job prioritization for pull-based job activation. Job priority affects the order in which jobs are activated for workers. It does not change retries, failures, or completion semantics.

Reviewer 2: Worth adding a sentence or two upfront on when you'd actually need this such as, "Use job prioritization when you have mixed-urgency workloads and want time-sensitive jobs to be processed ahead of lower-priority ones." Right now the section jumps straight into what it is, but readers landing here cold won't know if it's relevant to them.

You can define job priority:

- On the process as a default.
- On supported job-creating tasks as an override, including service tasks, send tasks, and script or business rule tasks implemented as job workers.

Important behavior:

> - If you do not specify job priorities, jobs keep the existing FIFO behavior (order by arrival/creation).

Reviewer 2: A few of these are saying the same thing in different ways. Items 3 and 4 are essentially both saying "ordering is per-partition, not global." Items 1 and 5/7 all touch on FIFO and default priority 0. Worth collapsing those into fewer, cleaner bullets or grouping them into small paragraphs by theme (scope of ordering, defaults, starvation). 9 items in a flat list is a lot to parse.

Suggested rewrite without the important behavior heading:

Job priority is enforced per partition, not globally across the cluster. Within a partition, higher-priority jobs are activated before lower-priority ones. Within the same priority level, jobs activate in FIFO order. If you don't set any priorities, standard FIFO behavior applies throughout.

The default priority is 0. A task-level priority definition overrides any process-level default.

Be aware that low-priority jobs can starve if workers are consistently occupied by higher-priority work. Camunda doesn't provide starvation mitigation, fairness guarantees, or priority aging. If starvation becomes a problem, increase worker capacity or revise your priority assignments.

- Higher-priority jobs are activated before lower-priority jobs within a partition.
- Ordering is best effort across the cluster, not a global guarantee across all partitions.
- Because jobs are distributed across partitions, priority ordering is applied per partition rather than as one globally sorted queue.
- If no priority is defined, the default priority is `0`.
- Task-level priority overrides process-level priority.
- Within the same priority, jobs are activated in FIFO order.
- Low-priority jobs can starve under sustained high-priority load and may never be activated while workers remain fully occupied by higher-priority jobs.
- Camunda does not provide fairness, aging, or starvation mitigation in the engine.

If starvation becomes a problem, increase worker capacity or revise your priority assignments.

Priority values can be static integers or FEEL expressions. FEEL expressions are evaluated when the job is created, and the resulting integer is stored on the job.

> Priority uses signed 32-bit integer semantics. Do not assume a fixed `0-99` engine limit.

Reviewer 2: Maybe this is just me cos the concept is quite new to me so this comment is optional and you can totally ignore my suggestion: 

This reads oddly to someone like me who isn't coming from a system that uses `0–99` priority scales. The assumption isn't self-evident. Maybe better as a positive statement: "Priority accepts any signed 32-bit integer. There's no fixed upper or lower bound enforced by the engine."

> :::warning Compatibility for pre-8.10 jobs

Reviewer 2: This callout is doing two different jobs at once. One is just context (execution defaults to priority 0, no action needed), the other is a genuine gotcha (API queries silently exclude pre-8.10 jobs). I'd split them: keep the execution behavior as plain prose and reserve the callout for the API filtering consequence, which is the part that can actually surprise someone.

Suggested rewrite:

Jobs created before version 8.10 do not have a stored `priority` value. For job execution,
Camunda treats them as having the default priority of `0`.
:::warning
If you use job APIs that filter or sort by `priority`, pre-8.10 jobs are excluded from
results because no priority value is stored for them. Review any priority-based queries
before upgrading.
:::

Jobs created before 8.10 do not have a populated `priority` value in secondary storage.

For job execution behavior, Camunda treats these jobs as if they had the default priority value of `0`.

However, if you use job APIs that filter or sort by `priority`, pre-8.10 jobs are filtered out because no priority value is stored for them.
:::

### Validation and runtime failure behavior

> Job priority values can be static integers or FEEL expressions.

Reviewer 2: This is a direct repeat of line 281. One of them needs to go.

Invalid FEEL expressions are handled differently depending on when they fail:

- If the expression is invalid at deployment time, BPMN deployment fails and the process is not deployed.
- If the expression fails at runtime, for example because a referenced variable is missing, Camunda raises an incident and the job is not created.

> Current recovery is manual. You can provide the missing variable and then resolve the incident.

Reviewer 2: ```suggestion
To recover, provide the missing variable value and resolve the incident manually in Operate.
```

### Job streaming limitation

Job streaming does not honor job priority ordering.

> If you mix job pull and job streaming, the benefits of prioritization can be reduced.

Reviewer 2: "The benefits of prioritization can be reduced" is a bit vague. Does it mean partially degraded, fully ignored, unpredictable? Worth being specific about what actually happens to priority-ordered jobs when some workers are on streaming. If the practical takeaway is "don't mix pull and streaming if priority ordering matters to you," that's worth stating directly.
